core/opendaq/reader/src/block_reader_impl.cpp

```diff
@@ -178,43 +178,51 @@ ErrCode BlockReaderImpl::readPacketData()
     return OPENDAQ_SUCCESS;
 }
 
-ErrCode BlockReaderImpl::readPackets(IReaderStatus** status)
+ErrCode BlockReaderImpl::readPackets(IReaderStatus** status, SizeT* count)
 {
     ErrCode errCode = OPENDAQ_SUCCESS;
 
     BlockReadInfo::Duration remainingTime = info.timeout;
-    auto shouldReadMore = getAvailable() > 0 || remainingTime.count() > 0;
+    auto availableBlocks = getAvailable();
+    const auto maxBlocksToRead = info.remainingSamplesToRead / blockSize;
 
-    while (info.remainingSamplesToRead > 0 && shouldReadMore)
+    while (availableBlocks < maxBlocksToRead && remainingTime.count() > 0)
+    {
+        // if not enough samples wait for the timeout or a full block
+
+        std::unique_lock notifyLock(notify.mutex);
+        if (notify.condition.wait_for(notifyLock, remainingTime, [this, &availableBlocks, maxBlocksToRead]
+        {
+             if (!notify.dataReady)
+                 return false;
+             availableBlocks = getAvailable();
+             notify.dataReady = false;
+             if (availableBlocks >= maxBlocksToRead)
+                return true;
+             return false;
+        }))
```

Comment:
I don't know if this is according to some guideline, but to me this is confusing as the {} block is a part of the lambda, not the if..true statement. I would indent the block to make it more obvious.